libs/mpi/include/hpx/mpi/mpi_future.hpp

```diff
@@ -0,0 +1,186 @@
+    hpx::future<void> async(F f, Ts &&...ts)
+    {
+        future_data_ptr data =
+            new future_data(future_data::init_no_addref{});
+
+        // invoke the call to MPI_Ixxx
+        f(std::forward<Ts>(ts)..., &data->request_);
+
+        // add the new shared state to the list for tracking
+        add_to_request_list(data);
+
+        // return a new future with the mpi::future_data shared state
+        using traits::future_access;
+        return future_access<hpx::future<void>>::create(std::move(data));
+    }
+
+    // -----------------------------------------------------------------
+    // Background progress function for MPI async operations
+    // Checks for completed MPI_Requests and sets mpi::future ready
+    // when found
+    void poll()
+    {
+        std::unique_lock<std::mutex> lk(list_mtx_, std::try_to_lock);
+        if (!lk.owns_lock()) return;
+
+        // for debugging
+        static auto timer = mpi_debug.make_timer(1, hpx::debug::str<>("MPI_Test"));
+        int rank = -1;
+        MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+        // list to track outstanding requests
+        // list useful to allow deleting from any location
+        std::list<future_data_ptr>::iterator i = active_requests_.begin();
+        while (i != active_requests_.end())
+        {
+            MPI_Request request = (*i)->request_;
+            int flag;
+            MPI_Status status;
+            //
+            auto original_req = request;
+            MPI_Test(&request, &flag, &status);
+            if (flag!=0) {
+                mpi_debug.debug(hpx::debug::str<>("MPI_Test set_data")
+                               , "R", rank
+                               , "request", hpx::debug::hex<8>(original_req)
+                               , "list size", active_requests_.size());
+
+                // mark the future as ready by setting the shared_state
+                (*i)->set_data(hpx::util::unused);
+                i = active_requests_.erase(i);
+            }
+            else {
+                mpi_debug.timed(timer
+                               , "no promise"
+                               , "R", rank
+                               , "request", hpx::debug::hex<8>(original_req)
+                               , "list size", active_requests_.size());
+            }
+            mpi_debug.timed(timer
+                           , "R", rank
+                           , "request", hpx::debug::hex<8>(original_req)
+                           , "list size", active_requests_.size());
+
+            ++i;
+        }
+    }
+
+    // This is not completely safe as it will return when the request list is
+    // empty, but cannot guarantee that other communications are not about
+    // to be launched in outstanding continuations etc.
+    void wait() {
+        // Yield until there is only this and background threads left.
+        hpx::util::yield_while([&]() {
+            return (active_requests_.size()>0);
+        });
```

Comment:
Would a `counting_semaphore` work here?

Reply:
Also, can this go into `namespace detail`?